numpy/core/src/multiarray/einsum.c.src

```diff
@@ -27,6 +27,15 @@
 #include "einsum_sumprod.h"
 #include "einsum_debug.h"
 
+static sum_of_products_fn
+get_sum_of_products_function(int nop, int type_num, npy_intp itemsize, npy_intp const *fixed_strides)
+{
+    #ifndef NPY_DISABLE_OPTIMIZATION
+        #include "einsum.dispatch.h"
+    #endif
```

Comment:
Does putting this include within the function body really work? If we don't already, we should have a CI test with this flag defined.

Reply:
this include is enabled by default, which is the entrance to all universal intrinsic code.

Reply:
> Does putting this include within the function body really work?

Yes, because this is an auto-generated config header that only contains two abstract macros used
by the dispatching macros to determine the required optimizations.

It's already included by `einsum_sumprod.h` but we tend to include it again since auto-generated config headers '*.dispatch.h' are overriding each other which allows the possibility of 'race condition' if another config header
has been involved in the scope.

However, we can move it before the function definition, not gonna make a difference.

Reply:
> If we don't already, we should have a CI test with this flag defined.

we have a test unit for the dispatching mechanism itself by it doesn't include this approach.

https://github.com/numpy/numpy/blob/5bf6d68e7241498556ee1b28fc146c1c21fbb5e4/numpy/core/src/umath/_umath_tests.c.src#L580-L591

Reply:
> It's already included by einsum_sumprod.h but we tend to include it again since auto-generated config headers '*.dispatch.h' are overriding each other

Can you elaborate on this? This sounds like a design flaw.

Reply:
> Can you elaborate on this?

I'm preparing a tutorial for CPU dispatcher contains an explanation for it. 

> This sounds like a design flaw

I wouldn't consider it a design flaw but a necessary deflection.

Reply:
Which things specifically "are overriding each other"?

Reply:
Before we go farther let me summarize the CPU dispatching process into three points:

1. Multiple compilations, requires **dispatch-able source** and macro `NPY_CPU_DISPATCH_CURFX`
2. Forward declarations, requires the auto-generated **dispatch-able configuration header** and usually macro `NPY_CPU_DISPATCH_DECLARE`
3. Runtime dispatching requires the auto-generated **dispatch-able configuration header** and usually macro `NPY_CPU_DISPATCH_CALL`

See [SIMD's Doc - dispatch-able sources](https://numpy.org/devdocs/reference/simd/simd-optimizations.html#dispatch-able-sources-and-configuration-statements) not completed yet but it contains a good explanation for dispatch-able source.

if you wanna dive more then you have to follow:

- the main auto-generated header `_cpu_dispatch.h` which contains all the #definitions and headers of instruction-sets
- auto-generated wrapped source `xxx.dispatch.x.c` where `xxx` is the dispatch-able source name and `x` is the required
  optimization.
- auto-generated dispatch-able configuration header `xxx.dispatch.h` where `xxx` is the dispatch-able source name.
- [npy_cpu_dispatch.h](https://github.com/numpy/numpy/blob/master/numpy/core/src/common/npy_cpu_dispatch.h) which defines  ``NPY_CPU_DISPATCH_`* macros.
- I wouldn't dare to ask you to check [`CCompilerOpt`](https://github.com/numpy/numpy/blob/5bf6d68e7241498556ee1b28fc146c1c21fbb5e4/numpy/distutils/ccompiler_opt.py) without improving doc string and finishing the main SIMD document,  but its the only way I guess to fully understanding how everything works.

---

Now back to your question,

>  Which things specifically "are overriding each other"?

`CCompilerOpt` generate a configuration header for each dispatch-able source contains two abstract config macros `NPY__CPU_DISPATCH_CALL`(repeater for dispatched features) and `NPY__CPU_DISPATCH_BASELINE_CALL`(single call if baseline option exists between). see this [comment](https://github.com/mattip/numpy/issues/46#issuecomment-678703443)

These two abstract macros are used by macros `NPY_CPU_DISPATCH_*` in `npy_cpu_dispatch.h` to determine the required optimizations that defined by configurations statements(`/*@targets*/`).  

If you looked inside the auto-generated **dispatch-able configuration header** you will see the header #undef any previous definitions
of  `NPY__CPU_DISPATCH_CALL` and `NPY__CPU_DISPATCH_BASELINE_CALL`, in other words each config header overriding each other.

Now assume you have multiple dispatch-able sources exporting(linking time) variant functions, each dispatch-able source
requires different optimizations through the configurations statements`(/*@targets*/)` and at the same time
you need to dispatch or declare these exported functions under the a scope of one source.

For example (forward declarations):

```C
#ifndef NPY_DISABLE_OPTIMIZATION
     #include "add_kernels.dispatch.h" 
#endif
// the following call will proceed multiple declarations according to the required optimizations in "add_kernels.dispatch.h"
NPY_CPU_DISPATCH_DECLARE(add_kernel, (params..))
#ifndef NPY_DISABLE_OPTIMIZATION
     #include "sub_kernels.dispatch.h"
#endif
// the following call will proceed multiple declarations according to the required optimizations in "sub_kernels.dispatch.h"
// since "sub_kernels.dispatch.h" #undef the previous abstract config macros of "add_kernels.dispatch.h
NPY_CPU_DISPATCH_DECLARE(sub_kernel, (params..))
```

The same example is applicable to runtime dispatching via macro `NPY_CPU_DISPATCH_CALL` .
In the light of the above, we require the including of  the auto-generated **dispatch-able configuration header** before any use of
`NPY_CPU_DISPATCH_*` macros to guarantee of having the right configuration header to avoid preprocessing race condition.

Reply:
> If you looked inside the auto-generated dispatch-able configuration header you will see the header #undef any previous definitions
of NPY__CPU_DISPATCH_CALL and NPY__CPU_DISPATCH_BASELINE_CALL, in other words each config header overriding each other.

Can you paste such a file in a gist somewhere?

Reply:
The main auto-generated header [_cpu_dispatch.h](https://gist.github.com/seiko2plus/6d8f1b08c62e485810f28d3b159a88a2) on x86 according to the default build settings.

The generated files that related to dispatch-able source [_umath_tests.dispatch.c](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/_umath_tests.dispatch.c) on x86 according to the default build settings:

- [the auto-generated wrapped source of avx2](https://gist.github.com/seiko2plus/c9d6e38bc2cfee3ee4a697e449481929)
- [the auto-generated wrapped source of sse41](https://gist.github.com/seiko2plus/41d2c907de7439022fd05b59547a882d)
- there's no wrapped source for the baseline, we pass the dispatch-able source as-is.
- [the auto-generated dispatch-able configuration header](https://gist.github.com/seiko2plus/f5c8c302c38c42401c430be74322202d)